### ISSUE TITLE
Cargo dependency cannot recognize `~` as home path in `dependency`'s `path` field

### ISSUE THREAD

**Issue by eval-exec** [rust-lang/cargo#14882]
### Problem

I have a dependency on my local file system:
```toml
❯ cat Cargo.toml
[package]
name = "cargo-path-bug"
version = "0.1.0"
edition = "2021"

[dependencies]
cargo-path-bug-lib = { path = "~/Temp/cargo-path-bug-lib/", package = "cargo-path-bug-lib"}
```
But when I run `cargo run`, it recognize `~` as a dir, not my home path.
```bash
❯ cargo run
error: failed to get `cargo-path-bug-lib` as a dependency of package `cargo-path-bug v0.1.0 (/home/exec/Temp/cargo-path-bug)`

Caused by:
  failed to load source for dependency `cargo-path-bug-lib`

Caused by:
  Unable to update /home/exec/Temp/cargo-path-bug/~/Temp/cargo-path-bug-lib

Caused by:
  failed to read `/home/exec/Temp/cargo-path-bug/~/Temp/cargo-path-bug-lib/Cargo.toml`

Caused by:
  No such file or directory (os error 2)
```

### Possible Solution(s)

request to recognize `~` as home path.

### Version

```text

```

**Comment by epage** [comment]
Could you help me understand why you want to use user-relative paths?  In my mind, the use case is limited because no one else can use that.  If someone clones your repo, they have to lay out the paths exactly how you do which is even weirder for windows systems because users don't normally write to `~`.  For local prototyping or patching, I would assume relative paths would work just as well.  At worse, you can fully specify the path.

Support for `~` is not built-in to the OS but is something apps have to manually support.  Technically, adding support would be a breaking change as `~` is allowed in paths.

**Comment by eval-exec** [comment]
> Could you help me understand why you want to use user-relative paths?

emmm.I just got used to doing it this way.

>  If someone clones your repo, they have to lay out the paths exactly how you do which is even weirder for windows systems because users don't normally write to ~.

I think windows systems have `~` too, it's `%USERPROFILE%`, may like `C:\Users\John`

**Comment by epage** [comment]
Yes, most software that supports `~` maps it to the user profile. I was more so highlighting that users don't normally put files there and so 
- Its less convenient as more nesting is needed
- Cloning a repo that uses `~` will require unusual filesystem layout for a windows user
